core/arch/arm/pta/benchmark.c

```diff
@@ -0,0 +1,199 @@
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <bench.h>
+#include <compiler.h>
+#include <kernel/misc.h>
+#include <kernel/mutex.h>
+#include <kernel/pseudo_ta.h>
+#include <malloc.h>
+#include <mm/core_memprot.h>
+#include <mm/tee_mm.h>
+#include <mm/tee_pager.h>
+#include <pta_benchmark.h>
+#include <string.h>
+#include <string_ext.h>
+#include <stdio.h>
+#include <trace.h>
+
+#define TA_NAME		"benchmark.ta"
+
+struct tee_ts_global *bench_ts_global;
+static struct mutex bench_reg_mu = MUTEX_INITIALIZER;
+
+static TEE_Result rpc_reg_global_buf(uint64_t type, paddr_t phta, size_t size)
+{
+	struct optee_msg_param rpc_params;
+
+	memset(&rpc_params, 0, sizeof(rpc_params));
+	rpc_params.attr = OPTEE_MSG_ATTR_TYPE_VALUE_INPUT;
+	rpc_params.u.value.a = type;
+	rpc_params.u.value.b = (uint64_t)phta;
+	rpc_params.u.value.c = size;
+
+	return thread_rpc_cmd(OPTEE_MSG_RPC_CMD_BENCH_REG, 1, &rpc_params);
+}
+
+static TEE_Result register_benchmark_memref(uint32_t type,
+				TEE_Param p[TEE_NUM_PARAMS])
+{
+	TEE_Result res;
+
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_MEMREF_INOUT) ||
+		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
+		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	/*
+	 * We accept only non-secure buffers, as we later perform
+	 * registration of this buffer in NS layers
+	 * (optee linux kmod/optee client)
+	 */
+	if (!tee_pbuf_is_non_sec(p[0].memref.buffer, p[0].memref.size))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mutex_lock(&bench_reg_mu);
+
+	DMSG("Registering timestamp buffer, addr = %p, paddr = %" PRIxPA "\n",
+			p[0].memref.buffer,
+			virt_to_phys(p[0].memref.buffer));
+	bench_ts_global = p[0].memref.buffer;
+
+	/* send back to the optee linux kernel module */
+	res = rpc_reg_global_buf(OPTEE_MSG_RPC_CMD_BENCH_REG_NEW,
+			virt_to_phys((void *)bench_ts_global),
+			sizeof(struct tee_ts_global) +
+			sizeof(struct tee_ts_cpu_buf) *
+			bench_ts_global->cores);
+
+	mutex_unlock(&bench_reg_mu);
```

Comment:
I was trying to handle the case, when `register_benchmark_memref` is called twice (for the case, when someone has executed `optee_benchmark` app at the same time), so if don't to wrap `rpc_reg_global_buf` with mutex, it's possible to have a race condition in `static void handle_rpc_func_cmd_bm_reg(struct optee_msg_arg *arg)` [L312](https://github.com/linaro-swg/linux/pull/38/files#diff-9b23115d70507633758c1768dfe4de1bR312)

I think it's better to add condition in the beginning of `register_benchmark_memref` which will check if we have registered a buffer before, and then we can avoid such situation and unlock mutex before rpc call